### PR TITLE
   mgr/volumes: Fix pending_subvolume_deletions in volume info

### PR DESCRIPTION
**Pull request by kotreshhr**
Problem:
     The pending_subvolume_deletions field in 'fs volume info' output is
    always zero
    
Cause:
     The 'os.listdir' being used to list the trash directory didn't have
    subvolume basdir context. Hence it was always failing with ENOENT.
    
Fix:
     Modify the fs_util's listdir function to list the files as well and use it.
    
Signed-off-by: Kotresh HR <khiremat@redhat.com>
Fixes: https://tracker.ceph.com/issues/62278

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [x] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: qa/tasks/cephfs/test_volumes.py [hunk 1]
```diff
@@ -649,6 +649,20 @@ def test_volume_info(self):
         self.assertEqual(vol_info["used_size"], 0,
                          "Size should be zero when volumes directory is empty")
 
+    def test_volume_info_pending_subvol_deletions(self):
+        """
+        Tests the pending_subvolume_deletions in 'fs volume info' command
+        """
+        subvolname = self._generate_random_subvolume_name()
+        # create subvolume
+        self._fs_cmd("subvolume", "create", self.volname, subvolname)
```
Comment (neesingh-rh) [inline qa/tasks/cephfs/test_volumes.py:658]: `subvolume create` needs `--mode` argument to be set, otherwise `_do_subvolume_io` won't execute

**Review by vshankar — APPROVED** [review]

File: qa/tasks/cephfs/test_volumes.py (hunk 1 above)
Comment (kotreshhr) [reply qa/tasks/cephfs/test_volumes.py]: Fixed it.

**Review by neesingh-rh — APPROVED** [review]

**Comment by kotreshhr** [comment]
jenkins test make check arm64

**Comment by vshankar** [comment]
Integration branch - https://shaman.ceph.com/builds/ceph/wip-vshankar-testing-20230808.093601/

Test runs in ~3 hrs.

**Comment by vshankar** [comment]
@kotreshhr - https://pulpito.ceph.com/vshankar-2023-08-08_12:05:39-fs:volumes-wip-vshankar-testing-20230808.093601-testing-default-smithi/

One failure - https://pulpito.ceph.com/vshankar-2023-08-08_12:05:39-fs:volumes-wip-vshankar-testing-20230808.093601-testing-default-smithi/7362773

Timeout out removing a volume:

```
2023-08-08T12:46:49.923 DEBUG:teuthology.orchestra.run.smithi012:> sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 120 ceph --cluster ceph fs volume rm volume_0000000000128263 --yes-i-really-mean-it
...
...
...
2023-08-08T12:48:49.979 DEBUG:teuthology.orchestra.run:got remote process result: 124

```

Looks unrelated, but can we get an RCA please.

**Comment by kotreshhr** [comment]
> @kotreshhr - https://pulpito.ceph.com/vshankar-2023-08-08_12:05:39-fs:volumes-wip-vshankar-testing-20230808.093601-testing-default-smithi/
> ... 
> Looks unrelated, but can we get an RCA please.

From the logs, I see that the 'fs volume rm' couldn't get the thread lock of  `class AsyncJobs` while cancelling the pending purge jobs on that volume.

```
  def delete_fs_volume(self, volname, confirm):
        ...
        self.purge_queue.cancel_jobs(volname)  <<<< Cancel pending jobs on the volume.
        self.connection_pool.del_connections(volname, wait=True)
        return delete_volume(self.mgr, volname, metadata_pool, data_pools)

class AsyncJobs(threading.Thread):
...
...
   def cancel_job(self, volname, job):
        with lock_timeout_log(self.lock):         <<<<< Waiting for lock                                  
            return self._cancel_job(volname, job)
```

Mgr logs: We can see the possible deadlock warning message as well in the below log.

```
023-08-08T12:46:50.315+0000 7f110f49e700  0 [volumes INFO volumes.module] Starting _cmd_fs_volume_rm(prefix:fs volume rm, target:['mon-mgr', ''], vol_name:volume_0000000000128263, yes-i-really-mean-it:--yes-i-really-mean-it) < ""
2023-08-08T12:46:50.319+0000 7f110f49e700  0 [volumes DEBUG root] mon_command: 'config get' -> 0 in 0.002s
2023-08-08T12:46:50.323+0000 7f110f49e700  0 [volumes DEBUG mgr_util] locking <locked _thread.lock object at 0x7f1131ea6e70> with 5 timeout
2023-08-08T12:46:55.322+0000 7f110f49e700  0 [volumes DEBUG mgr_util] locking <locked _thread.lock object at 0x7f1131ea6e70> with 5 timeout
2023-08-08T12:47:00.322+0000 7f110f49e700  0 [volumes DEBUG mgr_util] locking <locked _thread.lock object at 0x7f1131ea6e70> with 5 timeout
2023-08-08T12:47:05.322+0000 7f110f49e700  0 [volumes DEBUG mgr_util] locking <locked _thread.lock object at 0x7f1131ea6e70> with 5 timeout
2023-08-08T12:47:10.322+0000 7f110f49e700  0 [volumes DEBUG mgr_util] locking <locked _thread.lock object at 0x7f1131ea6e70> with 5 timeout
2023-08-08T12:47:11.350+0000 7f1115cab700  0 [volumes INFO mgr_util] scanning for idle connections..
2023-08-08T12:47:11.350+0000 7f1115cab700  0 [volumes INFO mgr_util] cleaning up connections: []
2023-08-08T12:47:11.358+0000 7f1112ca5700  0 [volumes INFO mgr_util] scanning for idle connections..
2023-08-08T12:47:11.358+0000 7f1112ca5700  0 [volumes DEBUG mgr_util] fs_name (cephfs) connections ([<mgr_util.CephfsConnectionPool.Connection object at 0x7f1131e9fa90>])
2023-08-08T12:47:11.358+0000 7f1112ca5700  0 [volumes INFO mgr_util] cleaning up connections: []
2023-08-08T12:47:11.382+0000 7f111a4b4700  0 [volumes INFO mgr_util] scanning for idle connections..
2023-08-08T12:47:11.382+0000 7f111a4b4700  0 [volumes DEBUG mgr_util] fs_name (cephfs) connections ([<mgr_util.CephfsConnectionPool.Connection object at 0x7f1131e9fd60>])
2023-08-08T12:47:11.382+0000 7f111a4b4700  0 [volumes INFO mgr_util] cleaning up connections: [('cephfs', <mgr_util.CephfsConnectionPool.Connection object at 0x7f1131e9fd60>)]
2023-08-08T12:47:11.382+0000 7f111a4b4700  0 [volumes DEBUG mgr_util] self.fs_id=8, fs_id=9
2023-08-08T12:47:11.382+0000 7f111a4b4700  0 [volumes INFO mgr_util] aborting connection from cephfs 'cephfs'
2023-08-08T12:47:11.382+0000 7f111a4b4700  0 [volumes INFO mgr_util] abort done from cephfs 'cephfs'
2023-08-08T12:47:15.322+0000 7f110f49e700  0 [volumes DEBUG mgr_util] locking <locked _thread.lock object at 0x7f1131ea6e70> with 5 timeout
2023-08-08T12:47:20.322+0000 7f110f49e700  0 [volumes INFO mgr_util] possible deadlock acquiring <locked _thread.lock object at 0x7f1131ea6e70>
2023-08-08T12:47:20.322+0000 7f110f49e700  0 [volumes DEBUG mgr_util] locking <locked _thread.lock object at 0x7f1131ea6e70> with 5 timeout
2023-08-08T12:47:25.322+0000 7f110f49e700  0 [volumes DEBUG mgr_util] locking <locked _thread.lock object at 0x7f1131ea6e70> with 5 timeout
2023-08-08T12:47:30.322+0000 7f110f49e700  0 [volumes DEBUG mgr_util] locking <locked _thread.lock object at 0x7f1131ea6e70> with 5 timeout
2023-08-08T12:47:35.322+0000 7f110f49e700  0 [volumes DEBUG mgr_util] locking <locked _thread.lock object at 0x7f1131ea6e70> with 5 timeout
2023-08-08T12:47:40.326+0000 7f110f49e700  0 [volumes DEBUG mgr_util] locking <locked _thread.lock object at 0x7f1131ea6e70> with 5 timeout
```

I looked  into the logs further. The last time the same thread lock was taken during the `subvolume rm` on `subvolume_0000000000552516` which is done as part of `test_volume_info_pending_subvol_deletions` which is the test added  in this PR.

mgr logs:
```
2023-08-08T12:44:03.382+0000 7f110f49e700  0 [volumes INFO volumes.module] Starting _cmd_fs_subvolume_rm(prefix:fs subvolume rm, sub_name:subvolume_0000000000552516, target:['mon-mgr', ''], vol_name:cephfs) < ""
2023-08-08T12:44:03.382+0000 7f110f49e700  0 [volumes DEBUG volumes.fs.operations.lock] entering global lock
2023-08-08T12:44:03.382+0000 7f110f49e700  0 [volumes DEBUG volumes.fs.operations.lock] acquired global lock
2023-08-08T12:44:03.382+0000 7f110f49e700  0 [volumes DEBUG mgr_util] [get] volume: (cephfs) connection: ([])
2023-08-08T12:44:03.382+0000 7f110f49e700  0 [volumes DEBUG mgr_util] [get] spawning new connection since no connection is unused and we still have room for more
2023-08-08T12:44:03.382+0000 7f110f49e700  0 [volumes DEBUG mgr_util] Connecting to cephfs 'cephfs'
2023-08-08T12:44:03.382+0000 7f110f49e700  0 [volumes DEBUG mgr_util] Setting user ID and group ID of CephFS mount as root...
2023-08-08T12:44:03.382+0000 7f110f49e700  0 [volumes DEBUG mgr_util] CephFS initializing...
2023-08-08T12:44:03.390+0000 7f110f49e700  0 [volumes DEBUG mgr_util] CephFS mounting...
2023-08-08T12:44:03.410+0000 7f110f49e700  0 [volumes DEBUG mgr_util] Connection to cephfs 'cephfs' complete
2023-08-08T12:44:03.414+0000 7f110f49e700  0 [volumes DEBUG volumes.fs.operations.versions.subvolume_base] loading config 'subvolume_0000000000552516' [mode: new]
2023-08-08T12:44:03.414+0000 7f110f49e700  0 [volumes DEBUG volumes.fs.operations.versions.subvolume_base] discovering subvolume 'subvolume_0000000000552516' [mode: new]
2023-08-08T12:44:03.414+0000 7f110f49e700  0 [volumes DEBUG volumes.fs.operations.versions.metadata_manager] opening config b'/volumes/_nogroup/subvolume_0000000000552516/.meta'
2023-08-08T12:44:03.422+0000 7f110f49e700  0 [volumes DEBUG volumes.fs.operations.versions.subvolume_base] loaded subvolume 'subvolume_0000000000552516'
2023-08-08T12:44:03.422+0000 7f110f49e700  0 [volumes DEBUG volumes.fs.operations.versions.subvolume_base] loading config 'subvolume_0000000000552516' [mode: new]
2023-08-08T12:44:03.422+0000 7f110f49e700  0 [volumes DEBUG volumes.fs.operations.versions.metadata_manager] opening config b'/volumes/_nogroup/subvolume_0000000000552516/.meta'
2023-08-08T12:44:03.422+0000 7f110f49e700  0 [volumes DEBUG volumes.fs.operations.versions.metadata_manager] opening config b'/volumes/_nogroup/subvolume_0000000000552516/.meta'
2023-08-08T12:44:03.426+0000 7f110f49e700  0 [volumes DEBUG volumes.fs.operations.versions.subvolume_v2] refreshed metadata, checking subvolume path 'b'/volumes/_nogroup/subvolume_0000000000552516/d181dc2f-5499-4fff-a3d5-1b4e3f6f9adb''
2023-08-08T12:44:03.442+0000 7f110f49e700  0 [volumes INFO volumes.fs.operations.versions.subvolume_base] subvolume path 'b'/volumes/_nogroup/subvolume_0000000000552516'' moved to trashcan
2023-08-08T12:44:03.446+0000 7f110f49e700  0 [volumes INFO volumes.fs.async_job] queuing job for volume 'cephfs'
2023-08-08T12:44:03.446+0000 7f110f49e700  0 [volumes DEBUG mgr_util] locking <unlocked _thread.lock object at 0x7f1131ea6e70> with 5 timeout
2023-08-08T12:44:03.446+0000 7f110f49e700  0 [volumes DEBUG mgr_util] locked <locked _thread.lock object at 0x7f1131ea6e70>
2023-08-08T12:44:03.446+0000 7f110f49e700  0 [volumes DEBUG mgr_util] [put] connection: <mgr_util.CephfsConnectionPool.Connection object at 0x7f1131e9fc10> usage: 1
2023-08-08T12:44:03.446+0000 7f110f49e700  0 [volumes DEBUG volumes.fs.operations.lock] exited global lock
2023-08-08T12:44:03.446+0000 7f110f49e700  0 [volumes INFO volumes.module] Finishing _cmd_fs_subvolume_rm(prefix:fs subvolume rm, sub_name:subvolume_0000000000552516, target:['mon-mgr', ''], vol_name:cephfs) < ""

```
teuthology logs:  See below the subvolume `subvolume_0000000000552516`  is part of the test added in this PR

```
...
2023-08-08T12:44:02.934 DEBUG:teuthology.orchestra.run.smithi012:> sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 120 ceph --cluster ceph fs subvolume rm cephfs subvolume_0000000000552516
2023-08-08T12:44:03.484 DEBUG:teuthology.orchestra.run.smithi012:> sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 120 ceph --cluster ceph fs volume info cephfs
2023-08-08T12:44:03.963 DEBUG:teuthology.orchestra.run.smithi012:> sudo sh -c 'iptables-save | grep -v teuthology | iptables-restore'
2023-08-08T12:44:04.053 DEBUG:teuthology.orchestra.run.smithi012:> set -ex
2023-08-08T12:44:04.054 DEBUG:teuthology.orchestra.run.smithi012:> dd if=/proc/self/mounts of=/dev/stdout
2023-08-08T12:44:04.070 DEBUG:teuthology.orchestra.run.smithi012:> sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 120 ceph --cluster ceph osd dump --format=json
2023-08-08T12:44:04.559 DEBUG:teuthology.orchestra.run.smithi012:> sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 120 ceph --cluster ceph osd pool get .mgr pg_num
2023-08-08T12:44:05.002 DEBUG:teuthology.orchestra.run.smithi012:> sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 120 ceph --cluster ceph osd pool get rbd pg_num
2023-08-08T12:44:05.526 DEBUG:teuthology.orchestra.run.smithi012:> sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 120 ceph --cluster ceph osd pool get cephfs_metadata pg_num
2023-08-08T12:44:05.984 DEBUG:teuthology.orchestra.run.smithi012:> sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 120 ceph --cluster ceph osd pool get cephfs_data pg_num
2023-08-08T12:44:06.435 DEBUG:teuthology.orchestra.run.smithi012:> sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 120 ceph --cluster ceph osd blocklist ls
2023-08-08T12:44:06.863 DEBUG:teuthology.orchestra.run.smithi012:> sudo fusermount -u /home/ubuntu/cephtest/mnt.0
2023-08-08T12:44:06.911 DEBUG:teuthology.orchestra.run.smithi012:> (cd /home/ubuntu/cephtest && exec rmdir -- /home/ubuntu/cephtest/mnt.0)
2023-08-08T12:44:07.112 DEBUG:teuthology.orchestra.run.smithi012:> sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 120 ceph --cluster ceph osd dump --format=json
2023-08-08T12:44:07.630 DEBUG:teuthology.orchestra.run.smithi012:> sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 120 ceph --cluster ceph osd pool get .mgr pg_num
2023-08-08T12:44:08.130 DEBUG:teuthology.orchestra.run.smithi012:> sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 120 ceph --cluster ceph osd pool get rbd pg_num
2023-08-08T12:44:08.600 DEBUG:teuthology.orchestra.run.smithi012:> sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 120 ceph --cluster ceph osd pool get cephfs_metadata pg_num
2023-08-08T12:44:09.026 DEBUG:teuthology.orchestra.run.smithi012:> sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 120 ceph --cluster ceph osd pool get cephfs_data pg_num
2023-08-08T12:44:09.451 DEBUG:teuthology.orchestra.run.smithi012:> sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 120 ceph --cluster ceph osd blocklist ls
2023-08-08T12:44:10.078 DEBUG:teuthology.orchestra.run.smithi012:> set -ex
2023-08-08T12:44:10.079 DEBUG:teuthology.orchestra.run.smithi012:> dd if=/proc/self/mounts of=/dev/stdout
2023-08-08T12:44:10.096 DEBUG:teuthology.orchestra.run.smithi012:> (cd /home/ubuntu/cephtest && exec rmdir -- /home/ubuntu/cephtest/mnt.2)
2023-08-08T12:44:10.241 DEBUG:teuthology.orchestra.run.smithi012:> sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 120 ceph --cluster ceph fs dump --format=json
2023-08-08T12:44:10.785 DEBUG:teuthology.orchestra.run.smithi012:> sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 120 ceph --cluster ceph osd dump --format=json
2023-08-08T12:44:11.215 DEBUG:teuthology.orchestra.run.smithi012:> sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 120 ceph --cluster ceph osd pool get .mgr pg_num
2023-08-08T12:44:11.664 DEBUG:teuthology.orchestra.run.smithi012:> sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 120 ceph --cluster ceph osd pool get rbd pg_num
2023-08-08T12:44:12.216 DEBUG:teuthology.orchestra.run.smithi012:> sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 120 ceph --cluster ceph osd pool get cephfs_metadata pg_num
2023-08-08T12:44:12.707 DEBUG:teuthology.orchestra.run.smithi012:> sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 120 ceph --cluster ceph osd pool get cephfs_data pg_num
2023-08-08T12:44:13.257 DEBUG:teuthology.orchestra.run.smithi012:> sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 120 ceph --cluster ceph fs dump --format=json
2023-08-08T12:44:13.873 DEBUG:teuthology.orchestra.run.smithi012:> sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 120 ceph --cluster ceph osd dump --format=json
2023-08-08T12:44:14.337 DEBUG:teuthology.orchestra.run.smithi012:> sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 120 ceph --cluster ceph fs dump --format=json
2023-08-08T12:44:14.821 DEBUG:teuthology.orchestra.run.smithi012:> sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 120 ceph --cluster ceph fs dump --format=json
2023-08-08T12:44:15.303 DEBUG:teuthology.orchestra.run.smithi012:> sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 120 ceph --cluster ceph osd dump --format=json
2023-08-08T12:44:15.736 DEBUG:teuthology.orchestra.run.smithi012:> sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 120 ceph --cluster ceph fs fail cephfs
2023-08-08T12:44:16.192 DEBUG:teuthology.orchestra.run.smithi012:> sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 120 ceph --cluster ceph fs rm cephfs --yes-i-really-mean-it
2023-08-08T12:44:17.200 DEBUG:teuthology.orchestra.run.smithi012:> sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 120 ceph --cluster ceph osd pool rm cephfs_metadata cephfs_metadata --yes-i-really-really
-mean-it
2023-08-08T12:44:18.508 DEBUG:teuthology.orchestra.run.smithi012:> sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 120 ceph --cluster ceph osd pool rm cephfs_data cephfs_data --yes-i-really-really-mean-it
2023-08-08T12:44:19.521 DEBUG:teuthology.orchestra.run.smithi012:> sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 120 ceph --cluster ceph config rm mon mon_allow_pool_delete
2023-08-08T12:44:19.950 DEBUG:teuthology.orchestra.run.smithi012:> sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 120 ceph --cluster ceph log 'Ended test tasks.cephfs.test_volumes.TestVolumes.test_volume
_info_pending_subvol_deletions'

```

We can also see from the mgr logs, after the `subvolume rm` is finished, the purge thread finishes the  `rmtree` and unregistered the job.

```
2023-08-08T12:44:03.522+0000 7f1110ca1700  0 [volumes DEBUG volumes.fs.async_job] processing 1 volume entries
2023-08-08T12:44:03.522+0000 7f1110ca1700  0 [volumes DEBUG volumes.fs.purge_queue] fetching trash entry for volume 'cephfs'
2023-08-08T12:44:03.522+0000 7f1110ca1700  0 [volumes DEBUG mgr_util] [get] volume: (cephfs) connection: ([<mgr_util.CephfsConnectionPool.Connection object at 0x7f1131e9fa90>, <mgr_util.CephfsConnectionPool.Connection object at 0x7f1131e9f6a0>])
2023-08-08T12:44:03.522+0000 7f1110ca1700  0 [volumes DEBUG mgr_util] [get] connection: <mgr_util.CephfsConnectionPool.Connection object at 0x7f1131e9fa90> usage: 1
2023-08-08T12:44:03.522+0000 7f1110ca1700  0 [volumes DEBUG mgr_util] [get] connection: <mgr_util.CephfsConnectionPool.Connection object at 0x7f1131e9f6a0> usage: 0
2023-08-08T12:44:03.522+0000 7f1110ca1700  0 [volumes DEBUG mgr_util] self.fs_id=5, fs_id=5
2023-08-08T12:44:03.522+0000 7f1110ca1700  0 [volumes DEBUG mgr_util] [get] connection (<mgr_util.CephfsConnectionPool.Connection object at 0x7f1131e9f6a0>) can be reused
2023-08-08T12:44:03.522+0000 7f1110ca1700  0 [volumes DEBUG mgr_util] [put] connection: <mgr_util.CephfsConnectionPool.Connection object at 0x7f1131e9f6a0> usage: 1
2023-08-08T12:44:03.914+0000 7f110f49e700  0 [volumes INFO volumes.module] Starting _cmd_fs_volume_info(prefix:fs volume info, target:['mon-mgr', ''], vol_name:cephfs) < ""
2023-08-08T12:44:03.914+0000 7f110f49e700  0 [volumes DEBUG volumes.fs.operations.lock] entering global lock
2023-08-08T12:44:03.914+0000 7f110f49e700  0 [volumes DEBUG volumes.fs.operations.lock] acquired global lock
2023-08-08T12:44:03.914+0000 7f110f49e700  0 [volumes DEBUG mgr_util] [get] volume: (cephfs) connection: ([<mgr_util.CephfsConnectionPool.Connection object at 0x7f1131e9fc10>])
2023-08-08T12:44:03.914+0000 7f110f49e700  0 [volumes DEBUG mgr_util] [get] connection: <mgr_util.CephfsConnectionPool.Connection object at 0x7f1131e9fc10> usage: 0
2023-08-08T12:44:03.914+0000 7f110f49e700  0 [volumes DEBUG mgr_util] self.fs_id=5, fs_id=5
2023-08-08T12:44:03.914+0000 7f110f49e700  0 [volumes DEBUG mgr_util] [get] connection (<mgr_util.CephfsConnectionPool.Connection object at 0x7f1131e9fc10>) can be reused
2023-08-08T12:44:03.926+0000 7f110f49e700  0 [volumes DEBUG mgr_util] [put] connection: <mgr_util.CephfsConnectionPool.Connection object at 0x7f1131e9fc10> usage: 1
2023-08-08T12:44:03.926+0000 7f110f49e700  0 [volumes DEBUG volumes.fs.operations.lock] exited global lock
2023-08-08T12:44:03.926+0000 7f110f49e700  0 [volumes INFO volumes.module] Finishing _cmd_fs_volume_info(prefix:fs volume info, target:['mon-mgr', ''], vol_name:cephfs) < ""
2023-08-08T12:44:11.346+0000 7f1115cab700  0 [volumes INFO mgr_util] scanning for idle connections..
2023-08-08T12:44:11.346+0000 7f1115cab700  0 [volumes INFO mgr_util] cleaning up connections: []
2023-08-08T12:44:11.354+0000 7f1112ca5700  0 [volumes INFO mgr_util] scanning for idle connections..
2023-08-08T12:44:11.354+0000 7f1112ca5700  0 [volumes DEBUG mgr_util] fs_name (cephfs) connections ([<mgr_util.CephfsConnectionPool.Connection object at 0x7f1131e9fa90>, <mgr_util.CephfsConnectionPool.Connection object at 0x7f1131e9f6a0>])
2023-08-08T12:44:11.354+0000 7f1112ca5700  0 [volumes INFO mgr_util] cleaning up connections: []
2023-08-08T12:44:11.378+0000 7f111a4b4700  0 [volumes INFO mgr_util] scanning for idle connections..
2023-08-08T12:44:11.378+0000 7f111a4b4700  0 [volumes DEBUG mgr_util] fs_name (cephfs) connections ([<mgr_util.CephfsConnectionPool.Connection object at 0x7f1131e9fc10>])
2023-08-08T12:44:11.378+0000 7f111a4b4700  0 [volumes INFO mgr_util] cleaning up connections: []
2023-08-08T12:44:15.578+0000 7f11124a4700  0 [volumes DEBUG mgr_util] [put] connection: <mgr_util.CephfsConnectionPool.Connection object at 0x7f1131e9fa90> usage: 1
2023-08-08T12:44:15.578+0000 7f11124a4700  0 [volumes DEBUG volumes.fs.async_job] unregistering async job cephfs.b'd2082c1f-02b9-4580-94b5-3af15fe79b15' from thread <JobThread(purgejob.0, started daemon 139711298029312)>
2023-08-08T12:44:16.578+0000 7f11124a4700  0 [volumes DEBUG volumes.fs.async_job] processing 1 volume entries
2023-08-08T12:44:16.578+0000 7f11124a4700  0 [volumes DEBUG volumes.fs.purge_queue] fetching trash entry for volume 'cephfs'

```


I  am not exactly sure what's happening here.  I also see the locking instrumentation added to root cause deadlocks is missing in `JobThread` which takes the same lock of `AsyncJobs`. I will send out a PR for that.

@vshankar Is it possible for this deadlock to happen when the previous test is not waiting for the subvolume data removal and went ahead with the new test ?

**Comment by vshankar** [comment]
> @vshankar Is it possible for this deadlock to happen when the previous test is not waiting for the subvolume data removal and went ahead with the new test ?

I don't see how that could happen. Two things
- Do you see any backtraces for async threads?
- The lock instrumentation code does not take into account the thread-id of the caller. Basically, if a thread acquires a lock and spends more than 5s doing work (with the lock acquired), another thread could trigger the "possible deadlock" warning.

**Comment by kotreshhr** [comment]
> > @vshankar Is it possible for this deadlock to happen when the previous test is not waiting for the subvolume data removal and went ahead with the new test ?
> 
> I don't see how that could happen. Two things
> 
>     * Do you see any backtraces for async threads?

No, there are no async threads backtraces.

> 
>     * The lock instrumentation code does not take into account the thread-id of the caller. Basically, if a thread acquires a lock and spends more than 5s doing work (with the lock acquired), another thread could trigger the "possible deadlock" warning.

Yeah, but that is very unlikely since the actual job is done outside the lock right?

**Comment by kotreshhr** [comment]
> > Yeah, but that is very unlikely since the actual job is done outside the lock right?
> 
> Right. I was just mentioning about the case where the possible deadlock is a red-herring. The async purge and clone happen outside the lock.

I think the test failed for the same reason as before here [1]. I have pushed the PR https://github.com/ceph/ceph/pull/52921 to log the missing mutexes. @vshankar Let me know your opinion.


[1] https://tracker.ceph.com/issues/49605#note-5

**Comment by vshankar** [comment]
> > > Yeah, but that is very unlikely since the actual job is done outside the lock right?
> > 
> > Right. I was just mentioning about the case where the possible deadlock is a red-herring. The async purge and clone happen outside the lock.
> 
> I think the test failed for the same reason as before here [1]. I have pushed the PR https://github.com/ceph/ceph/pull/52921 to log the missing mutexes. @vshankar Let me know your opinion.
> 
> 
> [1] https://tracker.ceph.com/issues/49605#note-5

I see what you mean. The earlier test did not wait for the subvolume to get purged and fs delete blocked some calls from volumes plugin (purge thread).

Adding a wait in the _earlier_ test would work around this. So, this isn't a deadlock in mgr/volumes till the fs gets removed and the purge thread gets unlucky. Can we add some kind of warning (maybe a cluster warning)?

**Comment by vshankar** [comment]
Spoke to @kotreshhr - merging this as per https://github.com/ceph/ceph/pull/52765#issuecomment-1673411937. @kotreshhr will send out a dox change that makes a note on this issue.

**Pull request merged**